# Dlaczego wasz model jest lepszy od innych? 


Otrzymaliśmy 92% trafności na danych testowych. Uważamy, że model jest wystarczająco dobry, żeby móc go zaimplementować

# Jak to sprawdziliście? 


Model przetrenowaliśmy na oczyszczonych danych (9826 pomiarów). Zawsze otrzymaliśmy skuteczność około 91%

# Dlaczego podjęliście takie a nie inne decyzje?


Chcieliśmy uprościć problem jak najbardziej. Dlatego podczas oczyszczania danych usunęliśmy wszystkie wiersze, które zawierały NaNy - błędne pomiary nie powinny być interpretowane przez model. Następnie wybraliśmy tylko wartości określone w kolumnach jako średnia (mean) lub odchylenie standardowe (std) - czyli najważniejsze dla nas wartości. Dzięki temu zmnieszyliśmy wymiarowość danych do 66 cech. Żeby sieć nie uczyła się też na danych, które zostały błędnie odczytane, obliczyliśmy dla każdej osoby (subject) w ramach każdej aktywności, wartości średnie dla wspomnianych wcześniej cech. Dało to nam 160 pomiarów treningowych. To wystarczyło sieci neuronowej na bardzo efektywną naukę.

Użyliśmy funkcji aktywacji softmax na końcu ze względu na jej charakter. Zwraca nam ona listę prawdopodobieństw, która sumuje się do 1. Można dzięki temu interpretować, które aktywności są najbardziej podobne do danego pomiaru, a które mniej

Kryteria oceny:
1. Czyszczenie danych i redukcja wymiarowości. (10 pkt)*
2. Dokładność modelu. (25 pkt)
3. Zrozumienie i interpretacja wyników. (35 pkt)*
4. Prezentacja i użyteczność biznesowa. (20 pkt)

## Wstępna analiza

Wczytujemy plik csv używając biblioteki pandas jako DataFrame, upewniamy się że widoczne będą wszystkie kolumny z danymi, analizujemy wielkość i naturę zbioru danych.

In [1]:
import pandas as pd
df = pd.read_csv('data/final_train.csv')
pd.set_option('display.max_columns', None)

In [2]:
df.shape

(11086, 573)

In [3]:
df.sample(6)

Unnamed: 0          Activity  angle(X,gravityMean)  \
5724         5724          STANDING             -1.572540   
10222       10222           SITTING             -1.572895   
3445         3445  WALKING_UPSTAIRS             -1.567051   
5359         5359            LAYING             -1.550297   
8471         8471  WALKING_UPSTAIRS             -1.566817   
7868         7868            LAYING             -1.548333   

       angle(Y,gravityMean)  angle(Z,gravityMean)  \
5724              -0.831131             -2.444402   
10222             -0.576135             -2.442052   
3445              -0.920945             -2.487852   
5359               0.502990             -2.453401   
8471              -0.927742             -2.488468   
7868              -0.139628             -2.380889   

       angle(tBodyAccJerkMean),gravityMean)  angle(tBodyAccMean,gravity)  \
5724                              -1.874955                     0.101185   
10222                             -1.859323                     0.042777   
3445                              -1.892574                          NaN   
5359                              -1.847900                     0.425301   
8471                              -1.806468                    -0.397123   
7868                              -1.915239                     0.067613   

       angle(tBodyGyroJerkMean,gravityMean)  angle(tBodyGyroMean,gravityMean)  \
5724                               0.407919                          7.861519   
10222                              0.359664                          4.901132   
3445                               1.897501                          8.025782   
5359                               0.158019                          4.173614   
8471                               0.142059                          8.185654   
7868                              -0.887342                          6.007361   

       energy-mean()  fBodyAcc-bandsEnergy()-1,16  \
5724      -12.039372                     3.622975   
10222     -11.919376                     3.623727   
3445       -5.615832                     2.916779   
5359      -10.721792                     3.624078   
8471      -12.011341                     2.854836   
7868      -11.901262                     3.621984   

       fBodyAcc-bandsEnergy()-1,16.1  fBodyAcc-bandsEnergy()-1,16.2  \
5724                        6.147065                       2.496757   
10222                       6.156904                       2.496749   
3445                        5.550833                            NaN   
5359                        6.159178                       2.495173   
8471                        5.514802                       2.712511   
7868                        6.158630                       2.495276   

       fBodyAcc-bandsEnergy()-1,24  fBodyAcc-bandsEnergy()-1,24.1  \
5724                      5.776975                       3.277515   
10222                     5.777300                       3.309608   
3445                      5.600484                       1.539643   
5359                      5.777389                       3.316258   
8471                      5.572680                       1.443210   
7868                      5.776843                       3.314383   

       fBodyAcc-bandsEnergy()-1,24.2  fBodyAcc-bandsEnergy()-1,8  \
5724                       -1.352874                   -1.537859   
10222                      -1.352873                   -1.537875   
3445                       -1.354918                   -1.393032   
5359                       -1.352859                   -1.537941   
8471                       -1.354888                   -1.389771   
7868                       -1.352860                   -1.537483   

       fBodyAcc-bandsEnergy()-1,8.1  fBodyAcc-bandsEnergy()-1,8.2  \
5724                       2.836371                      8.616621   
10222                      2.844150                      8.615789   
3445                       2.452088                      8.154137   

W trakcie analizy dochodzimy do kilku wniosków, mianowicie:

In [4]:
pd.unique(df['Activity'])

array(['STANDING', 'WALKING_UPSTAIRS', 'LAYING', 'WALKING', 'SITTING',
       'WALKING_DOWNSTAIRS'], dtype=object)

1. Mamy 6 aktywności, które nasz model ma rozpoznawać,

In [5]:
df.subject.nunique()

30

2. W badaniu brało udział 30 uczestników,

3. Do trenowania modelu klasyfikacji definitywnie nie potrzebujemy wszystkich cech. Skupiamy się na tych, które zawierają średnią i odchylenie standardowe.

## Oczyszczanie danych

Zanim zaczniemy czyścić zbiór, wyciągamy z niego dane o uczestnikach i aktywnościach, to przyda nam się potem.

In [6]:
subject = df['subject']
activity = df['Activity']

Na początku wybieramy wszystkie kolumny zawierające średnią lub odchylenie standardowe wszystkich z cech.

In [7]:
mean_cols = [col for col in df.columns if 'mean' in col or 'Mean' in col]
std_cols = [col for col in df.columns if 'std' in col or 'Std' in col]

df = pd.concat([df[std_cols], df[mean_cols]], axis = 1)
df.shape

(11086, 87)

In [8]:
573-87, (573-87)/573

(486, 0.8481675392670157)

W ten sposób udało nam się pozbyć 486 cech, co stanowi prawie 85% całego zbioru.

Aktualnie nasz zbiór wygląda tak:

In [9]:
df.sample(4)

fBodyAcc-std()-X  fBodyAcc-std()-Y  fBodyAcc-std()-Z  \
6646          -1.260664         -3.766730          3.381222   
10630          0.770361         -1.415377          4.154967   
7314          -0.284085         -2.882954          3.925613   
10569         -0.452969         -2.748999          3.701299   

       fBodyAccJerk-std()-X  fBodyAccJerk-std()-Y  fBodyAccJerk-std()-Z  \
6646              -6.546935              3.138447              1.136508   
10630             -4.065111              0.378903              2.323065   
7314              -5.261620              2.082163              1.978638   
10569             -5.083413              1.637649              1.768109   

       fBodyAccMag-std()  fBodyBodyAccJerkMag-std()  \
6646           -3.559139                  -2.788984   
10630          -1.052975                  -3.004470   
7314           -1.678040                  -2.903099   
10569          -2.221924                  -2.930308   

       fBodyBodyGyroJerkMag-std()  fBodyBodyGyroMag-std()  fBodyGyro-std()-X  \
6646                    -6.083270                4.651769           1.962489   
10630                   -6.023437                4.826976           3.253739   
7314                    -6.052033                4.746901           2.268739   
10569                   -6.035813                4.760419           2.513767   

       fBodyGyro-std()-Y  fBodyGyro-std()-Z  tBodyAcc-std()-X  \
6646            0.645767           3.615101          5.657498   
10630           0.803673           4.685226          6.335615   
7314            0.719943           4.209043          5.963459   
10569           0.734279           4.160601          5.931106   

       tBodyAcc-std()-Y  tBodyAcc-std()-Z  tBodyAccJerk-std()-X  \
6646           3.691739          1.614544              4.508855   
10630          3.874086          3.327267              6.283645   
7314           3.758300          2.803582              5.423837   
10569          3.775123          2.348509              5.492946   

       tBodyAccJerk-std()-Y  tBodyAccJerk-std()-Z  tBodyAccJerkMag-std()  \
6646              -3.227274              3.088312               3.444910   
10630             -0.630992              3.893686               2.363531   
7314              -2.310950              3.624854               2.927659   
10569             -1.719491              3.532739               2.787447   

       tBodyAccMag-std()  tBodyGyro-std()-X  tBodyGyro-std()-Y  \
6646           -2.607666           4.088879           1.409021   
10630           1.326861           5.299794           5.497756   
7314            0.216474           4.313949           3.498119   
10569          -0.380145           4.572921           3.870383   

       tBodyGyro-std()-Z  tBodyGyroJerk-std()-X  tBodyGyroJerk-std()-Y  \
6646            0.634290              -3.610477              -0.413330   
10630           1.771050              -4.121929              -0.582748   
7314            1.294625              -3.561727              -0.522402   
10569           1.296387              -3.787568              -0.531979   

       tBodyGyroJerk-std()-Z  tBodyGyroJerkMag-std()  tBodyGyroMag-std()  \
6646                0.855758               -2.251233            3.999339   
10630               0.852488               -2.635729            5.647951   
7314                0.853818               -2.481992            4.861632   
10569               0.853243               -2.533601            5.049976   

       tGravityAcc-std()-X  tGravityAcc-std()-Y  tGravityAcc-std()-Z  \
6646              3.627962             1.204948             3.943105   
10630             3.693231             1.416013             4.035716   
7314              3.674527             1.409776             4.087511   
10569             3.374093             0.969296             4.081077   

       tGravityAccMag-std()  angle(X,gravityMean)  angle(Y,gravityMean)  \
6646              -3.021398             -1.567078             -0.882981   
1

Dodatkowo zdecydowaliśmy się pozbyć kolumn, które zawierają słowa kluczowe: 'angle', 'energy' i 'meanFreq'

In [10]:
angle_cols = [col for col in df.columns if 'angle' in col or 'Angle' in col]
energy_cols = [col for col in df.columns if 'energy' in col or 'Energy' in col]
meanFreq_cols = [col for col in df.columns if 'meanFreq' in col]

drop_cols = angle_cols + energy_cols + meanFreq_cols
df = df.drop(drop_cols, axis=1)

In [11]:
df.shape

(11086, 66)

Ostatecznie nasz zbiór ma 66 cech.

Dodajemy kolumny uczestnika i aktywności.

In [12]:
df.insert(0, 'Activity', activity)
df.insert(0, 'Subject', subject)
df

Subject            Activity  fBodyAcc-std()-X  fBodyAcc-std()-Y  \
0           30            STANDING          0.756296         -1.586674   
1            9    WALKING_UPSTAIRS         -0.537596         -2.784092   
2           28              LAYING          0.755296         -1.283612   
3           25             WALKING          0.149588         -2.796041   
4           28              LAYING          0.755296         -1.283612   
...        ...                 ...               ...               ...   
11081        9    WALKING_UPSTAIRS         -0.537596         -2.784092   
11082       13              LAYING          0.634064         -1.573505   
11083       23    WALKING_UPSTAIRS         -0.455705         -3.214993   
11084       25  WALKING_DOWNSTAIRS         -0.630574         -3.224995   
11085       16             WALKING         -0.240327         -2.373576   

       fBodyAcc-std()-Z  fBodyAccJerk-std()-X  fBodyAccJerk-std()-Y  \
0              4.141758             -4.176057              0.585073   
1              3.677794             -5.817531              1.293797   
2              4.217478             -4.063730              0.303414   
3              3.922191             -4.776575              1.763192   
4              4.217478             -4.063730              0.303414   
...                 ...                   ...                   ...   
11081          3.677794             -5.817531              1.293797   
11082          4.180632             -4.123368              0.390518   
11083          3.537636             -4.946013              1.706158   
11084          3.941542             -4.899729              1.996001   
11085          3.734079             -5.133128              1.172635   

       fBodyAccJerk-std()-Z  fBodyAccMag-std()  fBodyBodyAccJerkMag-std()  \
0                  2.245282          -1.000675                  -2.979941   
1                  1.616815          -2.166122                  -2.906000   
2                  2.333802          -0.806943                  -3.011841   
3                  1.876670          -1.633627                  -2.920310   
4                  2.333802          -0.806943                  -3.011841   
...                     ...                ...                        ...   
11081              1.616815          -2.166122                  -2.906000   
11082              2.295930          -1.208579                  -2.996996   
11083              1.567060          -1.850257                  -2.931433   
11084              2.006284          -2.144305                  -2.929620   
11085              1.342477          -1.771676                  -2.914053   

       fBodyBodyGyroJerkMag-std()  fBodyBodyGyroMag-std()  fBodyGyro-std()-X  \
0                       -6.033122                4.781842           2.565681   
1                       -6.062388                4.653310           1.148811   
2                       -6.017739                4.845153           3.302522   
3                       -6.048324                4.737144           1.710101   
4                       -6.017739                4.845153           3.302522   
...                           ...                     ...                ...   
11081                   -6.062388                4.653310           1.148811   
11082                   -6.034841                4.727674           3.014504   
11083                   -6.049794                4.689007           2.196733   
11084                   -6.062599                4.582817           0.636642   
11085                   -6.053385                4.772722           2.302898   

       fBodyGyro-std()-Y  fBodyGyro-std()-Z  tBodyAcc-std()-X  \
0               0.795967           4.632040          6.325173   
1               0.660377           3.794268          5.880477   
2               0.807615           4.705391          6.331799   
3               0.746790           4.400935          6.119588   
4               0.807615           4.705391          6

Ostatecznie zdecydowaliśmy się zgrupować wszystkich uczestników badania razem z aktywnościami, w których brali udział

In [13]:
grouped = df.groupby(['Subject', 'Activity'])

Wyciągamy średnią z wszystkich cech, dla każdego uczestnika względem każdej z badanych aktywności.

In [14]:
df = grouped.mean().reset_index()

I tak prezentują się nasze oczyszczone dane:

In [15]:
df.head(10)

Subject            Activity  fBodyAcc-std()-X  fBodyAcc-std()-Y  \
0        1              LAYING          0.742534         -1.352281   
1        1             SITTING          0.792379         -1.317112   
2        1            STANDING          0.802141         -1.296172   
3        1             WALKING         -0.421209         -3.346157   
4        1  WALKING_DOWNSTAIRS         -0.937546         -3.193000   
5        1    WALKING_UPSTAIRS         -0.360645         -3.236741   
6        2              LAYING          0.776581         -1.255021   
7        2             SITTING          0.786274         -1.333269   
8        2            STANDING          0.789142         -1.323450   
9        2             WALKING         -0.157225         -2.903061   

   fBodyAcc-std()-Z  fBodyAccJerk-std()-X  fBodyAccJerk-std()-Y  \
0          4.165208             -4.048067              0.307285   
1          4.209662             -4.026415              0.266434   
2          4.213388             -4.021199              0.269555   
3          3.827163             -5.742295              2.730468   
4          3.840883             -5.901125              2.468137   
5          3.630124             -5.110440              1.608362   
6          4.219369             -4.029950              0.263865   
7          4.201622             -4.031829              0.279734   
8          4.194748             -4.050078              0.305840   
9          3.922206             -5.412628              2.499345   

   fBodyAccJerk-std()-Z  fBodyAccMag-std()  fBodyBodyAccJerkMag-std()  \
0              2.339434          -0.946739                  -3.012469   
1              2.362534          -0.825279                  -3.015763   
2              2.364546          -0.804977                  -3.016560   
3              1.611067          -2.143042                  -2.846935   
4              1.413464          -2.660484                  -2.838354   
5              1.934546          -2.053104                  -2.928445   
6              2.361921          -0.784797                  -3.016474   
7              2.358947          -0.862932                  -3.015247   
8              2.349733          -0.855420                  -3.012713   
9              1.743108          -1.745344                  -2.858880   

   fBodyBodyGyroJerkMag-std()  fBodyBodyGyroMag-std()  fBodyGyro-std()-X  \
0                   -6.019987                4.825410           3.137130   
1                   -6.015538                4.833118           3.293136   
2                   -6.015802                4.848482           3.320864   
3                   -6.101496                4.649844           1.875883   
4                   -6.110421                4.662521           1.519395   
5                   -6.056959                4.586221           1.865585   
6                   -6.015841                4.843543           3.327022   
7                   -6.016409                4.842478           3.317201   
8                   -6.018685                4.841217           3.271087   
9                   -6.079439                4.746878           2.088899   

   fBodyGyro-std()-Y  fBodyGyro-std()-Z  tBodyAcc-std()-X  tBodyAcc-std()-Y  \
0           0.810149           4.673888          6.328449          3.878952   
1           0.810766           4.661738          6.344424          3.881932   
2           0.812218           4.703371          6.347711          3.883188   
3           0.654746           4.139156          5.908111          3.725957   
4           0.678570           3.915523          5.740868          3.731538   
5           0.624657           4.246821          5.960151          3.744934   
6           0.811642           4.682635          6.339698          3.885891   
7           0.810710           4.683007          6.342081          3.880483   
8           0.809159           4.688821          6.341898          3.880879   
9           0.736979           4.253549          5.996609      